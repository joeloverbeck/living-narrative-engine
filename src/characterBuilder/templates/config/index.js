/**
 * @file Template configuration system exports
 * @module characterBuilder/templates/config
 */

import { DEFAULT_TEMPLATE_CONFIGS, getDefaultConfig, mergeWithDefaults } from './defaultConfigs.js';
import { TemplateConfigManager } from '../utilities/templateConfigManager.js';
import { EnvironmentConfigLoader } from './environmentConfig.js';
import { ConfigValidator } from './configValidator.js';
import { TemplateConfigBuilder } from './configBuilder.js';
import { PageConfigRegistry } from './pageConfigs.js';

export {
  DEFAULT_TEMPLATE_CONFIGS,
  getDefaultConfig,
  mergeWithDefaults,
  TemplateConfigManager,
  EnvironmentConfigLoader,
  ConfigValidator,
  TemplateConfigBuilder,
  PageConfigRegistry,
};

// Re-export error classes
export {
  TemplateConfigurationError,
  InvalidConfigError,
  MissingConfigError,
  EnvironmentDetectionError,
  ConfigMergeError,
} from '../errors/templateConfigurationError.js';

/**
 * Create a fully configured template configuration system
 * 
 * @param {object} [options] - System options
 * @returns {object} Configuration system components
 */
export function createConfigurationSystem(options = {}) {
  const {
    enableCache = true,
    forceEnvironment = null,
    validateOnBuild = true,
  } = options;

  // Create environment loader
  const environmentLoader = new EnvironmentConfigLoader({
    forceEnvironment,
  });

  // Create validator
  const validator = new ConfigValidator();

  // Create config manager
  const configManager = new TemplateConfigManager({
    defaults: DEFAULT_TEMPLATE_CONFIGS,
    environment: environmentLoader.getEnvironment(),
    validator,
    enableCache,
  });

  // Create page registry
  const pageRegistry = new PageConfigRegistry();

  // Apply environment configs to manager
  const envConfig = environmentLoader.getConfig();
  for (const [templateType, config] of Object.entries(envConfig)) {
    configManager.setConfig('environment', templateType, config);
  }

  // Apply page configs to manager
  for (const pageName of pageRegistry.getRegisteredPages()) {
    const pageConfig = pageRegistry.getPageConfig(pageName);
    configManager.setConfig('page', pageName, pageConfig);
  }

  return {
    configManager,
    environmentLoader,
    validator,
    pageRegistry,
    createBuilder: (templateType) => new TemplateConfigBuilder({}, templateType, validateOnBuild),
  };
}