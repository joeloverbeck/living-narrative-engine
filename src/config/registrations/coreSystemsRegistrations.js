// src/core/config/registrations/coreSystemsRegistrations.js
// --- FILE START ---

/**
 * @fileoverview Registers core game logic systems, particularly those needing initialization or shutdown.
 * Also includes registration for turn-related handlers, resolvers, and the ITurnContext provider.
 */

// --- JSDoc Imports for Type Hinting ---
/** @typedef {import('../appContainer.js').default} AppContainer */
/** @typedef {import('../../interfaces/coreServices.js').ILogger} ILogger */
/** @typedef {import('../../turns/interfaces/ITurnManager.js').ITurnManager} ITurnManager */
/** @typedef {import('../../turns/interfaces/ITurnOrderService.js').ITurnOrderService} ITurnOrderService */
/** @typedef {import('../../interfaces/IValidatedEventDispatcher.js').IValidatedEventDispatcher} IValidatedEventDispatcher */
/** @typedef {import('../../interfaces/ISafeEventDispatcher.js').ISafeEventDispatcher} ISafeEventDispatcher */
/** @typedef {import('../../interfaces/eventBus.js').EventBus} EventBus */
/** @typedef {import('../../interfaces/IWorldContext.js').IWorldContext} IWorldContext */
/** @typedef {import('../../interfaces/IActionExecutor.js').IActionExecutor} IActionExecutor */
/** @typedef {import('../../services/actionValidationService.js').ActionValidationService} ActionValidationService */
/** @typedef {import('../../interfaces/IActionDiscoverySystem.js').IActionDiscoverySystem} IActionDiscoverySystem_Interface */ // MODIFIED: Changed name for clarity
/** @typedef {import('../../commands/interfaces/ICommandProcessor.js').ICommandProcessor} ICommandProcessor */
/** @typedef {import('../../turns/ports/IPromptOutputPort.js').IPromptOutputPort} IPromptOutputPort */
/** @typedef {import('../../turns/ports/ITurnEndPort.js').ITurnEndPort} ITurnEndPort */
/** @typedef {import('../../turns/handlers/playerTurnHandler.js').default} PlayerTurnHandler_Concrete */ // Renamed for clarity
/** @typedef {import('../../turns/handlers/aiTurnHandler.js').default} AITurnHandler_Concrete */       // Renamed for clarity
/** @typedef {import('../../turns/services/turnHandlerResolver.js').default} TurnHandlerResolver_Concrete */ // Renamed for clarity
/** @typedef {import('../../interfaces/IEntityManager.js').IEntityManager} IEntityManager_Interface */ // MODIFIED: Changed name for clarity
/** @typedef {import('../../interfaces/IGameDataRepository.js').IGameDataRepository} IGameDataRepository */
/** @typedef {import('../../turns/interfaces/IPlayerPromptService.js').IPlayerPromptService} IPlayerPromptService */
/** @typedef {import('../../commands/interfaces/ICommandOutcomeInterpreter.js').ICommandOutcomeInterpreter} ICommandOutcomeInterpreter */
/** @typedef {import('../../interfaces/ICommandInputPort.js').ICommandInputPort} ICommandInputPort */
/** @typedef {import('../../services/subscriptionLifecycleManager.js').default} SubscriptionLifecycleManager */
/** @typedef {import('../../turns/interfaces/ITurnHandler.js').ITurnHandler} ITurnHandler */
/** @typedef {import('../../turns/interfaces/ITurnContext.js').ITurnContext} ITurnContext */
/** @typedef {import('../../turns/interfaces/ILLMAdapter.js').ILLMAdapter} ILLMAdapter_Interface */
/** @typedef {import('../../interfaces/coreServices.js').ISchemaValidator} ISchemaValidator_Interface */ // <<< ADDED THIS LINE

// --- System Imports ---
import {ActionDiscoverySystem} from '../../systems/actionDiscoverySystem.js';
import TurnManager from '../../turns/turnManager.js';

// --- Handler & Resolver Imports ---
import PlayerTurnHandler from '../../turns/handlers/playerTurnHandler.js'; // Concrete class for factories
import AITurnHandler from '../../turns/handlers/aiTurnHandler.js';       // Concrete class for factories
import TurnHandlerResolver from '../../turns/services/turnHandlerResolver.js'; // Concrete Class

// --- DI & Helper Imports ---
import {tokens} from '../tokens.js';
import {Registrar} from '../registrarHelpers.js';
import {formatActionCommand} from '../../services/actionFormatter.js';
import {INITIALIZABLE, SHUTDOWNABLE} from "../tags.js";

/**
 * Registers core game systems tagged as initializable and/or shutdownable,
 * along with essential turn-handling services including the ITurnContext provider.
 * Excludes quest-related systems which are registered separately.
 *
 * @export
 * @param {AppContainer} container - The application's DI container.
 */
export function registerCoreSystems(container) {
    const registrar = new Registrar(container);
    /** @type {ILogger} */
    const logger = container.resolve(tokens.ILogger);

    let registrationCount = 0;
    logger.info('Core Systems Registration: Starting...');

    // --- Register ActionDiscoverySystem against its Interface Token using singletonFactory ---
    registrar.tagged(INITIALIZABLE).singletonFactory(tokens.IActionDiscoverySystem, (c) => new ActionDiscoverySystem({
        gameDataRepository: /** @type {IGameDataRepository} */ (c.resolve(tokens.IGameDataRepository)),
        entityManager: /** @type {IEntityManager_Interface} */ (c.resolve(tokens.IEntityManager)),
        actionValidationService: c.resolve(tokens.ActionValidationService),
        logger: c.resolve(tokens.ILogger),
        formatActionCommandFn: formatActionCommand,
        getEntityIdsForScopesFn: () => new Set()
    }));
    logger.debug(`Core Systems Registration: Registered ${String(tokens.IActionDiscoverySystem)} tagged with ${INITIALIZABLE.join(', ')}.`);
    registrationCount++;


    // --- Turn Handlers, Resolver, and Manager ---
    // PlayerTurnHandler registration remains for potential direct use if needed.
    registrar.tagged(SHUTDOWNABLE).singletonFactory(tokens.PlayerTurnHandler, (c) =>
        new PlayerTurnHandler({
            logger: /** @type {ILogger} */ (c.resolve(tokens.ILogger)),
            commandProcessor: /** @type {ICommandProcessor} */ (c.resolve(tokens.ICommandProcessor)),
            turnEndPort: /** @type {ITurnEndPort} */ (c.resolve(tokens.ITurnEndPort)),
            playerPromptService: /** @type {IPlayerPromptService} */ (c.resolve(tokens.IPlayerPromptService)),
            commandOutcomeInterpreter: /** @type {ICommandOutcomeInterpreter} */ (c.resolve(tokens.ICommandOutcomeInterpreter)),
            safeEventDispatcher: /** @type {ISafeEventDispatcher} */ (c.resolve(tokens.ISafeEventDispatcher)),
            subscriptionLifecycleManager: /** @type {SubscriptionLifecycleManager} */ (c.resolve(tokens.SubscriptionLifecycleManager))
        })
    );
    logger.debug(`Core Systems Registration: Registered ${tokens.PlayerTurnHandler} tagged ${SHUTDOWNABLE.join(', ')}.`);
    registrationCount++;

    // AITurnHandler registration is primarily for the factory now, but can remain for direct resolution if necessary.
    // If AITurnHandler itself needs to be SHUTDOWNABLE, add the tag here.
    registrar.singletonFactory(tokens.AITurnHandler, (c) => new AITurnHandler({
        logger: /** @type {ILogger} */ (c.resolve(tokens.ILogger)),
        gameWorldAccess: /** @type {IWorldContext} */ (c.resolve(tokens.IWorldContext)),
        turnEndPort: /** @type {ITurnEndPort} */ (c.resolve(tokens.ITurnEndPort)),
        illmAdapter: /** @type {ILLMAdapter_Interface} */ (c.resolve(tokens.ILLMAdapter)),
        commandProcessor: /** @type {ICommandProcessor} */ (c.resolve(tokens.ICommandProcessor)),
        commandOutcomeInterpreter: /** @type {ICommandOutcomeInterpreter} */ (c.resolve(tokens.ICommandOutcomeInterpreter)),
        safeEventDispatcher: /** @type {ISafeEventDispatcher} */ (c.resolve(tokens.ISafeEventDispatcher)),
        subscriptionManager: /** @type {SubscriptionLifecycleManager} */ (c.resolve(tokens.SubscriptionLifecycleManager)),
        entityManager: /** @type {IEntityManager_Interface} */ (c.resolve(tokens.IEntityManager)),
        actionDiscoverySystem: /** @type {IActionDiscoverySystem_Interface} */ (c.resolve(tokens.IActionDiscoverySystem)),
        schemaValidator: /** @type {ISchemaValidator_Interface} */ (c.resolve(tokens.ISchemaValidator)), // <<< MODIFIED: Added ISchemaValidator
    }));
    logger.debug(`Core Systems Registration: Registered ${tokens.AITurnHandler}.`);
    registrationCount++;

    // ****** MODIFIED SECTION START ******
    registrar.singletonFactory(tokens.TurnHandlerResolver, (c) => {
        // Define factory functions that will create NEW instances of handlers
        const createPlayerHandlerFactory = () => new PlayerTurnHandler({
            logger: /** @type {ILogger} */ (c.resolve(tokens.ILogger)),
            commandProcessor: /** @type {ICommandProcessor} */ (c.resolve(tokens.ICommandProcessor)),
            turnEndPort: /** @type {ITurnEndPort} */ (c.resolve(tokens.ITurnEndPort)),
            playerPromptService: /** @type {IPlayerPromptService} */ (c.resolve(tokens.IPlayerPromptService)),
            commandOutcomeInterpreter: /** @type {ICommandOutcomeInterpreter} */ (c.resolve(tokens.ICommandOutcomeInterpreter)),
            safeEventDispatcher: /** @type {ISafeEventDispatcher} */ (c.resolve(tokens.ISafeEventDispatcher)),
            subscriptionLifecycleManager: /** @type {SubscriptionLifecycleManager} */ (c.resolve(tokens.SubscriptionLifecycleManager)),
        });

        const createAiHandlerFactory = () => {
            const resolvedLogger = /** @type {ILogger} */ (c.resolve(tokens.ILogger));
            const resolvedIllmAdapter = /** @type {ILLMAdapter_Interface} */ (c.resolve(tokens.ILLMAdapter));
            const resolvedEntityManager = /** @type {IEntityManager_Interface} */ (c.resolve(tokens.IEntityManager));
            const resolvedActionDiscoverySystem = /** @type {IActionDiscoverySystem_Interface} */ (c.resolve(tokens.IActionDiscoverySystem));
            const resolvedSchemaValidator = /** @type {ISchemaValidator_Interface} */ (c.resolve(tokens.ISchemaValidator)); // <<< MODIFIED: Resolve ISchemaValidator

            if (!resolvedIllmAdapter) {
                resolvedLogger.error(`CoreSystemsRegistration: Failed to resolve ${String(tokens.ILLMAdapter)} for AITurnHandler factory.`);
                throw new Error(`Missing dependency ${String(tokens.ILLMAdapter)} for AITurnHandler factory`);
            }
            if (!resolvedEntityManager) {
                resolvedLogger.error(`CoreSystemsRegistration: Failed to resolve ${String(tokens.IEntityManager)} for AITurnHandler factory.`);
                throw new Error(`Missing dependency ${String(tokens.IEntityManager)} for AITurnHandler factory`);
            }
            if (!resolvedActionDiscoverySystem) {
                resolvedLogger.error(`CoreSystemsRegistration: Failed to resolve ${String(tokens.IActionDiscoverySystem)} for AITurnHandler factory.`);
                throw new Error(`Missing dependency ${String(tokens.IActionDiscoverySystem)} for AITurnHandler factory`);
            }
            if (!resolvedSchemaValidator) { // <<< MODIFIED: Add check for ISchemaValidator
                resolvedLogger.error(`CoreSystemsRegistration: Failed to resolve ${String(tokens.ISchemaValidator)} for AITurnHandler factory.`);
                throw new Error(`Missing dependency ${String(tokens.ISchemaValidator)} for AITurnHandler factory`);
            }
            return new AITurnHandler({
                logger: resolvedLogger,
                gameWorldAccess: /** @type {IWorldContext} */ (c.resolve(tokens.IWorldContext)),
                turnEndPort: /** @type {ITurnEndPort} */ (c.resolve(tokens.ITurnEndPort)),
                illmAdapter: resolvedIllmAdapter,
                commandProcessor: /** @type {ICommandProcessor} */ (c.resolve(tokens.ICommandProcessor)),
                commandOutcomeInterpreter: /** @type {ICommandOutcomeInterpreter} */ (c.resolve(tokens.ICommandOutcomeInterpreter)),
                safeEventDispatcher: /** @type {ISafeEventDispatcher} */ (c.resolve(tokens.ISafeEventDispatcher)),
                subscriptionManager: /** @type {SubscriptionLifecycleManager} */ (c.resolve(tokens.SubscriptionLifecycleManager)),
                entityManager: resolvedEntityManager,
                actionDiscoverySystem: resolvedActionDiscoverySystem,
                schemaValidator: resolvedSchemaValidator, // <<< MODIFIED: Pass ISchemaValidator
            });
        };

        return new TurnHandlerResolver({
            logger: /** @type {ILogger} */ (c.resolve(tokens.ILogger)),
            createPlayerTurnHandler: createPlayerHandlerFactory,
            createAiTurnHandler: createAiHandlerFactory
        });
    });
    // ****** MODIFIED SECTION END ******
    logger.debug(`Core Systems Registration: Registered ${tokens.TurnHandlerResolver} (with handler factories).`);
    registrationCount++;

    registrar.tagged(INITIALIZABLE).singletonFactory(tokens.ITurnManager, (c) => new TurnManager({
        turnOrderService: /** @type {ITurnOrderService} */ (c.resolve(tokens.ITurnOrderService)),
        entityManager: /** @type {IEntityManager_Interface} */ (c.resolve(tokens.IEntityManager)),
        logger: /** @type {ILogger} */ (c.resolve(tokens.ILogger)),
        dispatcher: /** @type {IValidatedEventDispatcher} */ (c.resolve(tokens.IValidatedEventDispatcher)),
        turnHandlerResolver: /** @type {TurnHandlerResolver_Concrete} */ (c.resolve(tokens.TurnHandlerResolver))
    }));
    logger.debug(`Core Systems Registration: Registered ${tokens.ITurnManager} tagged ${INITIALIZABLE.join(', ')}.`);
    registrationCount++;

    container.register(
        tokens.ITurnContext,
        (c) => {
            const localLogger = /** @type {ILogger} */ (c.resolve(tokens.ILogger));
            const turnManager = /** @type {ITurnManager | null} */ (c.resolve(tokens.ITurnManager));

            if (!turnManager) {
                localLogger.warn(`ITurnContext Factory: ${String(tokens.ITurnManager)} could not be resolved. Returning null.`);
                return null;
            }

            const activeHandler = turnManager.getActiveTurnHandler();
            if (activeHandler && typeof activeHandler.getTurnContext === 'function') {
                const context = activeHandler.getTurnContext();
                return context;
            } else if (activeHandler) {
                localLogger.warn(`ITurnContext Factory: Active handler (${activeHandler.constructor.name}) found, but getTurnContext is not a function. Returning null.`);
            }
            return null;
        },
        {lifecycle: 'transient'}
    );
    logger.debug(`Core Systems Registration: Registered transient factory for ${String(tokens.ITurnContext)}.`);
    registrationCount++;


    logger.info(`Core Systems Registration: Completed registering ${registrationCount} systems, handlers, services, and providers.`);
}

// --- FILE END ---