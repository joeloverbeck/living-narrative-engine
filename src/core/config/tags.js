// src/core/tags.js

/**
 * Array containing the tag used to identify systems that require initialization.
 * @constant
 * @type {string[]}
 */
export const INITIALIZABLE = ['initializableSystem'];
export const SHUTDOWNABLE = ['shutdownable']