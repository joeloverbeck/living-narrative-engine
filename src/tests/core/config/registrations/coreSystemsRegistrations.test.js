// src/core/config/registrations/coreSystemsRegistrations.test.js

// --- JSDoc Imports for Type Hinting ---
/** @typedef {import('../../../../core/interfaces/coreServices.js').ILogger} ILogger */
/** @typedef {any} AppContainer */ // Using 'any' as the custom container type isn't defined

// --- Jest Imports ---
import {describe, beforeEach, it, expect, jest} from '@jest/globals';

// --- Class Under Test ---
import {registerCoreSystems} from '../../../../core/config/registrations/coreSystemsRegistrations.js'; // Adjust path as needed

// --- Dependencies ---
import {tokens} from '../../../../core/config/tokens.js';
import {INITIALIZABLE} from "../../../../core/config/tags";

// --- Mock Implementations (Core Services) ---
const mockLogger = {
    info: jest.fn(), warn: jest.fn(), error: jest.fn(), debug: jest.fn(),
};

// --- Mock Custom DI Container (Simplified version for this test) ---
const createMockContainer = () => {
    const registrations = new Map();
    const container = {
        _registrations: registrations, // Expose for snapshot testing
        register: jest.fn((token, factoryOrValue, options = {}) => {
            if (!token) throw new Error('Mock Register Error: Token is required.');
            registrations.set(token, {factoryOrValue, options});
        }),
        resolve: jest.fn((token) => {
            // For this test, we only *really* need to resolve the logger
            if (token === tokens.ILogger) return mockLogger;
            // For other dependencies needed by system *factories* (if they were executed),
            // we'd need a more robust mock resolve. But here we just test registration calls.
            const registration = registrations.get(token);
            if (!registration) {
                // Allow resolving known tokens even if not strictly registered in *this* test
                // This helps if a factory tries to resolve something basic.
                // Example: If MoveCoordinatorSystem factory *was* executed and needed BlockerSystem
                if (token === tokens.EventBus) return {on: jest.fn(), off: jest.fn(), emit: jest.fn()};
                if (token === tokens.EntityManager) return { /* basic mock */};
                if (token === tokens.BlockerSystem) return { /* basic mock */};
                if (token === tokens.MovementSystem) return { /* basic mock */};
                if (token === tokens.GameDataRepository) return { /* basic mock */};
                if (token === tokens.GameStateManager) return { /* basic mock */};
                if (token === tokens.ActionExecutor) return { /* basic mock */};
                if (token === tokens.ConditionEvaluationService) return { /* basic mock */};
                if (token === tokens.ItemTargetResolverService) return { /* basic mock */};
                if (token === tokens.ValidatedEventDispatcher) return { /* basic mock */};
                if (token === tokens.ActionValidationService) return { /* basic mock */};

                // If it's none of the above known dependencies, it's likely an error in the test or SUT
                throw new Error(`Mock Resolve Error: Token not registered or mocked: ${String(token)}`);
            }
            // Basic resolve for testing purposes if needed, won't handle factories correctly here.
            if (typeof registration.factoryOrValue === 'function' && registration.options?.lifecycle === 'singleton') {
                // Return the factory itself for inspection, rather than executing it
                return registration.factoryOrValue;
            }
            return registration.factoryOrValue;
        }),
        // Add a mock resolveAll for SystemInitializer if it were tested here
        resolveAll: jest.fn((tag) => {
            const found = [];
            for (const [token, reg] of registrations.entries()) {
                if (reg.options?.tags?.includes(tag)) {
                    // In a real test needing instances, resolve here. For now, just return token.
                    found.push(token);
                }
            }
            return found;
        })
    };
    return container;
};

describe('registerCoreSystems', () => {
    /** @type {ReturnType<typeof createMockContainer>} */
    let mockContainer;

    // Whitelist of system tokens expected to be registered and tagged
    // UPDATED: Removed tokens.QuestSystem, tokens.QuestStartTriggerSystem
    const initializableSystemTokens = [
        tokens.GameRuleSystem, tokens.EquipmentEffectSystem, tokens.EquipmentSlotSystem,
        tokens.InventorySystem, tokens.CombatSystem, tokens.DeathSystem,
        tokens.WorldPresenceSystem, tokens.ItemUsageSystem, tokens.BlockerSystem,
        tokens.MovementSystem, tokens.MoveCoordinatorSystem,
        tokens.PerceptionSystem, tokens.NotificationUISystem,
        tokens.OpenableSystem, tokens.HealthSystem, tokens.StatusEffectSystem,
        tokens.LockSystem, tokens.ActionDiscoverySystem
    ];
    const expectedCount = 19;

    beforeEach(() => {
        jest.clearAllMocks();
        mockContainer = createMockContainer();

        // Pre-register the essential dependency needed *by* registerCoreSystems itself: ILogger
        mockContainer.register(tokens.ILogger, mockLogger); // No lifecycle needed for this mock setup
    });

    it(`should register ${expectedCount} systems/services with the '${INITIALIZABLE[0]}' tag`, () => { // <-- Test description updates automatically
        // Arrange
        registerCoreSystems(mockContainer);

        // Assert: Check that each whitelisted token was registered...
        initializableSystemTokens.forEach(token => {
            // ... this loop now won't check for WelcomeMessageService
            expect(mockContainer.register).toHaveBeenCalledWith(
                token,
                expect.any(Function),
                expect.objectContaining({
                    tags: expect.arrayContaining(INITIALIZABLE),
                    lifecycle: 'singleton'
                })
            );
        });

        expect(mockContainer.register).toHaveBeenCalledTimes(expectedCount);

        // Assert: Check logger calls
        expect(mockLogger.debug).toHaveBeenCalledWith('Core Systems Registration: Starting...');
        initializableSystemTokens.forEach(token => { // <-- This loop also won't check for WelcomeMessageService log
            expect(mockLogger.debug).toHaveBeenCalledWith(expect.stringContaining(`Registered ${String(token)} tagged with ${INITIALIZABLE[0]}`));
        });
        // Check final info log message count - should match the updated expectedCount
        // NOTE: You might need to update the count *inside* registerCoreSystems.js's final log message too if it's hardcoded!
        expect(mockLogger.info).toHaveBeenCalledWith(expect.stringContaining(`Core Systems Registration: Completed registering ${expectedCount} systems`));
    });

    it('should match snapshot for registration calls', () => {
        // Arrange
        registerCoreSystems(mockContainer);

        // Assert
        // Snapshot the calls made to the mock container's register function
        // Exclude the pre-registered mockLogger call for a cleaner snapshot focused on the bundle's work
        const callsToSnapshot = mockContainer.register.mock.calls.filter(
            (call) => call[0] !== tokens.ILogger
        );
        // NOTE: This snapshot will differ from the original due to the removal
        // of QuestSystem and QuestStartTriggerSystem registrations.
        // You will need to update the snapshot (`jest --updateSnapshot`).
        expect(callsToSnapshot).toMatchSnapshot();
    });
});