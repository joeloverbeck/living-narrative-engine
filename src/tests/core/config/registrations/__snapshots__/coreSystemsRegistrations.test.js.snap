// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerCoreSystems should match snapshot for registration calls 1`] = `
[
  [
    "EquipmentEffectSystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
        "GameDataRepository",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
      ],
    },
  ],
  [
    "EquipmentSlotSystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
        "GameDataRepository",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
      ],
    },
  ],
  [
    "InventorySystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
        "GameDataRepository",
        "IWorldContext",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
      ],
    },
  ],
  [
    "DeathSystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
      ],
    },
  ],
  [
    "ItemUsageSystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
        "ConditionEvaluationService",
        "ItemTargetResolverService",
        "GameDataRepository",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
      ],
    },
  ],
  [
    "MovementSystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
      ],
    },
  ],
  [
    "CombatSystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
        "GameDataRepository",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
        "shutdownable",
      ],
    },
  ],
  [
    "WorldPresenceSystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
        "shutdownable",
      ],
    },
  ],
  [
    "NotificationUISystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "GameDataRepository",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
        "shutdownable",
      ],
    },
  ],
  [
    "OpenableSystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
        "shutdownable",
      ],
    },
  ],
  [
    "HealthSystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
        "GameDataRepository",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
        "shutdownable",
      ],
    },
  ],
  [
    "StatusEffectSystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
        "GameDataRepository",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
        "shutdownable",
      ],
    },
  ],
  [
    "LockSystem",
    [Function],
    {
      "dependencies": [
        "EventBus",
        "EntityManager",
      ],
      "lifecycle": "singleton",
      "tags": [
        "initializableSystem",
        "shutdownable",
      ],
    },
  ],
  [
    "IActionDiscoverySystem",
    [Function],
    {
      "lifecycle": "singletonFactory",
      "tags": [
        "initializableSystem",
      ],
    },
  ],
  [
    "PlayerTurnHandler",
    [Function],
    {
      "lifecycle": "singletonFactory",
      "tags": [
        "shutdownable",
      ],
    },
  ],
  [
    "AITurnHandler",
    [Function],
    {
      "lifecycle": "singletonFactory",
    },
  ],
  [
    "TurnHandlerResolver",
    [Function],
    {
      "lifecycle": "singletonFactory",
    },
  ],
  [
    "ITurnManager",
    [Function],
    {
      "lifecycle": "singletonFactory",
      "tags": [
        "initializableSystem",
      ],
    },
  ],
]
`;
